File: erts-pc/ERTSDashboard/Communication/Data/RemoteControlData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ERTS.Dashboard.Communication.Data
{
    public class RemoteControlData : PackageData
    {
        ushort lift; //Throttle
        short roll; //Aileron
        short pitch; //Elevator
        short yaw; //Rudder
        
        /// <summary>
        /// The Throttle control signal
        /// </summary>
        public ushort Lift {
            get { return lift; }
        }
        /// <summary>
        /// The Aileron control signal
        /// </summary>
        public short Roll {
            get { return roll; }
        }
        /// <summary>
        /// The Elevator control signal
        /// </summary>
        public short Pitch {
            get { return pitch; }
        }
        /// <summary>
        /// The Rudder control signal
        /// </summary>
        public short Yaw {
            get { return yaw; }
        }
        public RemoteControlData(ushort Lift, short Roll, short Pitch, short Yaw)
        {
            lift = Lift;
            roll = Roll;
            pitch = Pitch;
            yaw = Yaw;
        }

        public RemoteControlData(byte[] data)
        {
            if (data.Length < Length)
                throw new ArgumentException("Data is not long enough.", "data");

            lift = BitConverter.ToUInt16(data, 0);
            roll = BitConverter.ToInt16(data, sizeof(ushort) * 1);
            pitch = BitConverter.ToInt16(data, sizeof(ushort) * 2);
            yaw = BitConverter.ToInt16(data, sizeof(ushort) * 3);
        }

        public RemoteControlData()
        {
            throw new NotSupportedException();
        }

        public override int Length => 4 * sizeof(ushort);

        public override bool ExpectsAcknowledgement => (false);

        public override bool IsValid()
        {
            return true;
        }

        public override byte[] ToByteArray()
        {

            byte[] data = new byte[Length];
            Buffer.BlockCopy(BitConverter.GetBytes(lift), 0, data, sizeof(ushort) * 0, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(roll), 0, data, sizeof(ushort) * 1, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(pitch), 0, data, sizeof(ushort) * 2, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(yaw), 0, data, sizeof(ushort) * 3, sizeof(ushort));
            return data;
        }

        public override void SetAckNumber(uint AckNumber)
        {
            throw new NotImplementedException();
        }
    }
}

File: libraries/libertscommunication/remote_control_data.cpp
#include "remote_control_data.h"

RemoteControlData::RemoteControlData(uint16_t lift,int16_t roll,int16_t pitch,int16_t yaw) {
	_lift = lift;
	_roll = roll;
	_pitch = pitch;
	_yaw = yaw;    
}
RemoteControlData::RemoteControlData(const uint8_t *data) {
	_lift = *(reinterpret_cast<const uint16_t*>(&data[0]));
	_roll = *(reinterpret_cast<const int16_t*>(&data[2]));
	_pitch = *(reinterpret_cast<const int16_t*>(&data[4]));
	_yaw = *(reinterpret_cast<const int16_t*>(&data[6]));
}

void RemoteControlData::to_buffer(uint8_t *buffer) {
	*(reinterpret_cast<uint16_t*>(&buffer[0])) = _lift;
	*(reinterpret_cast<int16_t*>(&buffer[2])) = _roll;
	*(reinterpret_cast<int16_t*>(&buffer[4])) = _pitch;
	*(reinterpret_cast<int16_t*>(&buffer[6])) = _yaw;
}
